### PR TITLE
bazel: 0.10.0 cc_configure.bzl support.

### PR DESCRIPTION
**Pull request by htuch**
Appears that the 0.10.0 release is not compatible with our cc_configure.bzl fork. This PR bumps our cc_configure.bzl to the 0.10.0 version
with the various Envoy-specific changes ported across.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

**Comment by htuch** [comment]
I haven't actually managed to fully test this; the 0.10.0 bump has exposed other issues with Bazel's include path validation and our toolchain with aliased paths. I need to upgrade my distribution to do this, ETA ~1.5 hours until I can do that. So, putting this up to allow others to validate in the interim.

You will at a minimum need to `bazel clean --expunge` (I nuked the full Bazel cache).

**Comment by mattklein123** [comment]
It's pretty hard to review this. Should this pass tests? Or do we have to force a new build image to get new bazel?

**Comment by htuch** [comment]
I'm surprised it's passing tests, given that CROSSTOOLS files are not backwards compatible in general. Two possible reasons are (1) we happened to get lucky and this CROSSTOOLs format change is compatible (2) when we `bazel fetch` in CI, it's already populated the CROSSTOOL file and isn't picking up the change here.

I need to validate this fix first locally. If it happens to work on CI and locally, I'd suggesting committing this PR and then doing an async bump of CI to 0.10.0. 

What could be done to make this easier to review?

**Comment by mattklein123** [comment]
> What could be done to make this easier to review?

I think I just need to look at the full new diffs, not the diff of diffs, which is making my head explode. I will do it when I get to the office.

File: bazel/cc_configure.bzl [hunk 1]
```diff
@@ -21,28 +21,36 @@
...
-  repository_ctx.file("tools/cpp/empty.cc", "int main() {}")
+def cc_autoconf_impl(repository_ctx, overriden_tools = dict()):
   repository_ctx.symlink(
       Label("@bazel_tools//tools/cpp:dummy_toolchain.bzl"), "dummy_toolchain.bzl")
+  env = repository_ctx.os.environ
   cpu_value = get_cpu_value(repository_ctx)
   if cpu_value == "freebsd":
-    fail("freebsd support needs to be added to the Envoy Bazel cc_configure.bzl fork")
+    # This is defaulting to the static crosstool, we should eventually
+    # autoconfigure this platform too.  Theorically, FreeBSD should be
```
Comment (jmarantz) [inline bazel/cc_configure.bzl]: Theoretically

File: bazel/cc_configure.bzl [hunk 2]
```diff
@@ -21,28 +21,36 @@
...
+    # straightforward to add but we cannot run it in a docker container so
```
Comment (jmarantz) [inline bazel/cc_configure.bzl]: Maybe add some commas

File: bazel/cc_configure.bzl [hunk 3]
```diff
@@ -21,28 +21,36 @@
...
+    # skipping until we have proper tests for FreeBSD.
+    repository_ctx.symlink(Label("@bazel_tools//tools/cpp:CROSSTOOL"), "CROSSTOOL")
+    repository_ctx.symlink(Label("@bazel_tools//tools/cpp:BUILD.static"), "BUILD")
   elif cpu_value == "x64_windows":
-    #configure_windows_toolchain(repository_ctx)
-    fail("x64_windows support needs to be added to the Envoy Bazel cc_configure.bzl fork")
-  elif cpu_value == "darwin":
-    configure_osx_toolchain(repository_ctx)
+    # TODO(ibiryukov): overriden_tools are only supported in configure_unix_toolchain.
+    # We might want to add that to Windows too(at least for msys toolchain).
```
Comment (jmarantz) [inline bazel/cc_configure.bzl]: space before paren.

File: bazel/unix_cc_configure.bzl [hunk 4]
```diff
@@ -158,10 +182,31 @@ def _is_gold_supported(repository_ctx, cc):
...
+def _get_no_canonical_prefixes_opt(repository_ctx, cc):
+  # If the compiler sometimes rewrites paths in the .d files without symlinks
+  # (ie when they're shorter), it confuses Bazel's logic for verifying all
+  # #included header files are listed as inputs to the action.
+
+  # The '-fno-canonical-system-headers' should be enough, but clang does not
+  # support it, so we also try '-no-canonical-prefixes' if first option does
+  # not work.
+  opt = _add_option_if_supported(repository_ctx, cc,
+                                 "-fno-canonical-system-headers")
```
Comment (jmarantz) [inline bazel/unix_cc_configure.bzl]: does this fit on one line?  What line-length limit are we going for in this file?  Here & 2 lines below.

**Comment by htuch** [comment]
Validate locally, will update CI later today.

**Pull request merged**